### PR TITLE
Use recursive algorithm when inserting networks

### LINKED ISSUE #69: IPv4 parent record and alias records are immutable
Right now it is possible to remove the IPv4 parent record or alias records by removing or replacing the subtree that they are in. This can lead to the unintentional removal of an IPv4 alias or, worse, leave the aliases with a dangling pointer to the IPv4 range. This is unlikely to happen when merging is enabled unless the subtree is explicitly removed or `force_overwrite` is used, but it would be very easy to accidentally do this when merging is disabled.

I'd propose that some records be marked as immutable, including all alias records and the IPv4 range. When removing or replacing a subtree, these records would remain unchanged (although the records in nodes that they point to may be updated).

### PR DESCRIPTION
**Pull request by oschwald**
This simplifies some of the logic and speeds up some inserts. The memory handling is also somewhat simplified and easier to keep track of.

Also, closes #69

### CODE REVIEW

File: t/MaxMind/DB/Writer/Tree-ipv6-aliases.t [hunk 1]
```diff
@@ -54,8 +54,8 @@ my $tempdir = tempdir( CLEANUP => 1 );
     # exception is someone tries to overwrite one. In the future, we could do
     # something smarter, _but_ it isn't clear what the right behavior is.
     like(
-        exception { $tree->insert_network( '2000::/4', {} ) },
-        qr/Attempted to free an IPv4 alias node./,
+        exception { $tree->insert_network( '2001::/32', {} ) },
+        qr/Attempted to overwrite an alised network./,
```
Comment (autarch) [inline t/MaxMind/DB/Writer/Tree-ipv6-aliases.t]: There's a typo here - "alised" - and in the corresponding error-throwing code.

File: c/tree.c [hunk 2]
```diff
@@ -535,24 +545,25 @@ LOCAL void resolve_ip(int tree_ip_version, const char *const ipstr,
...
 }
 
-struct network {
+typedef struct network {
     const char *const ipstr;
     const uint8_t prefix_length;
-};
+} network;
 
 static struct network ipv4_aliases[] = {
```
Comment (autarch) [inline c/tree.c]: Might as well use the typedef'd name here.

File: c/tree.c [hunk 3]
```diff
@@ -574,51 +585,51 @@ LOCAL void alias_ipv4_networks(MMDBW_tree_s *tree)
...
-                                                  NULL);
     free_network(&ipv4_root_network);
-    if (status != MMDBW_SUCCESS) {
-        croak("Unable to find IPv4 root node when setting up aliases");
-    }
 
-    if (MMDBW_RECORD_TYPE_NODE != ipv4_root_record->type) {
-        croak("Unexpected type for IPv4 root record: %s",
-              record_type_name(ipv4_root_record->type));
+    if (status != MMDBW_SUCCESS) {
```
Comment (autarch) [inline c/tree.c]: Most the code puts the macro first (which I did to protect against `=` vs `==` errors)

File: c/tree.c [hunk 4]
```diff
@@ -629,37 +640,190 @@ LOCAL MMDBW_status insert_record_for_network(
...
+        }
+    case MMDBW_RECORD_TYPE_FIXED_NODE:
+    case MMDBW_RECORD_TYPE_NODE: {
+            next_node = current_record->value.node;
+            break;
+        }
+    }
+
+    // If we are inserting a fixed node or an alias, we make all of the nodes
+    // down to that record fixed nodes. This makes it easier to not accidently
```
Comment (autarch) [inline c/tree.c]: Typo in comment

File: c/tree.c [hunk 5]
```diff
@@ -629,37 +640,190 @@ LOCAL MMDBW_status insert_record_for_network(
...
+        MMDBW_status status = insert_record_into_next_node(
+            tree,
+            &(next_node->right_record),
+            network,
+            current_bit + 1,
+            new_record,
+            merge_strategy,
+            is_internal_insert
+            );
+        if (status != MMDBW_SUCCESS) {
```
Comment (autarch) [inline c/tree.c]: Comparison order.

File: c/tree.c [hunk 6]
```diff
@@ -629,37 +640,190 @@ LOCAL MMDBW_status insert_record_for_network(
...
+        MMDBW_status status = insert_record_into_next_node(
+            tree,
+            &(next_node->left_record),
+            network,
+            current_bit + 1,
+            new_record,
+            merge_strategy,
+            is_internal_insert
+            );
         if (status != MMDBW_SUCCESS) {
```
Comment (autarch) [inline c/tree.c]: Comparison order.

File: c/tree.c (hunk 4 above)
Comment (oschwald) [inline c/tree.c]: If you are referring to "accidently", [Webster disagrees](https://www.merriam-webster.com/dictionary/accidental). That said, I fixed it.

**Pull request merged**